apm-sniffer/apm-sdk-plugin/resteasy-plugin/resteasy-server-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/resteasy/v3/server/SynchronousDispatcherInterceptor.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.resteasy.v3.server;
+
+import org.apache.skywalking.apm.agent.core.context.CarrierItem;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.tag.Tags;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+import org.jboss.resteasy.spi.HttpRequest;
+import org.jboss.resteasy.spi.HttpResponse;
+
+import java.lang.reflect.Method;
+
+/**
+ * @author yan-fucheng
+ */
+public class SynchronousDispatcherInterceptor implements InstanceMethodsAroundInterceptor {
+
+    @Override
+    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
+        MethodInterceptResult result) throws Throwable {
+        HttpRequest request = (HttpRequest) allArguments[0];
+
+        ContextCarrier contextCarrier = new ContextCarrier();
+        CarrierItem next = contextCarrier.items();
+        while (next.hasNext()) {
+            next = next.next();
+            next.setHeadValue(request.getHttpHeaders().getHeaderString(next.getHeadKey()));
+        }
+
+        AbstractSpan span = ContextManager.createEntrySpan(request.getUri().getPath(), contextCarrier);
+        Tags.URL.set(span, request.getUri().getRequestUri().toString());
+        Tags.HTTP.METHOD.set(span, request.getHttpMethod());
+        span.setComponent(ComponentsDefine.RESTEASY);
+        SpanLayer.asHttp(span);
+        objInst.setSkyWalkingDynamicField(span);
```

Comment:
If this request runs in sync mode, does it necessary to `#setSkyWalkingDynamicField`, because I noticed, all places of using `#getSkyWalkingDynamicField` are async interceptors, at the same time, `ContextManager.activeSpan().prepareForAsync();` called in another place. 

Is this the only way you could make plugin works?

In my mind, I prefer only set dynamic field if necessary, and make sure the `ContextManager.activeSpan().prepareForAsync();` result is set into the dynamic field(at the same place, if possible)

I am not familiar with RESTEasy lib, so I could be wrong. I just comment from the perspective of improving codes readability.

Reply:
Is there any way I can get the span directly from the `AsynchronousResponseInjectorInterceptor#beforeMethod` EnhancedInstance? How can I pass it through the dynamic field? I didn't figure it out, are there any tips?

Reply:
I am thinking about is this the right case of using asyncFinish APIs. In most case, when callback is async, we usually create a snapshot and start a new segment for the callback. Because, in some scenarios, people have further spans in callback, such as accessing the database.

`#asyncFinish` API is designed for vert.x originaly, for making sure the span duration is right. Is that your design?

Reply:
And also, my intention of this review is making sure the plugin is right. You save the span in sync(`SynchronousDispatcherInterceptor`) interceptor for the potential AsynchronousResponse.

Reply:
Again, I may be wrong. I am just talking based on your interceptor name and code styles. Please correct me if I am wrong.

Reply:
The resteasy async request separates the request and response in different threads. At this point, the connection is still waiting for this response. I think this should also be included in the span finsh time. If the snapshot finsh time is used, it may be 0ms. So I used `prepareForAsync()`, I don't know if it is the same as vert.x, I don't know if I am doing this right?

Reply:
If you think this response time makes sense, then I am good with that. I just remind you that, people could do further works in response callback(Am I right?), so even you use async finish to make the duration longer, still recommend you to do capture/continued.

Reply:
![image](https://user-images.githubusercontent.com/10338055/58087822-ed4c9800-7bf3-11e9-886e-9bfa584db366.png)

I have a problem. In this case, I don't want to enhance Runnable. Can I directly let the user solve it by using the toolkit?

Is there still a better way? I have no idea now.

Reply:
`Don't enhance Runnable` is 100% a good idea. Don't do that. If the user wants a runnable, let him do it by himself.

Reply:
I modified it and deleted the extra async part. waiting for test case.